### PR TITLE
Add Schema Directive

### LINKED ISSUE #134: Schema and Operations directives
Directives on schema doesn't seems to be executed.

* **Tartiflette version:** 0.4.0
* **Python version:** 3.7.1
* **Executed in docker:** No
* **Is a regression from a previous versions?** No

SDL example:
```graphql
directive @non_introspectable on SCHEMA | FIELD_DEFINITION

type Query {
  aField: String
}

schema @nonIntrospectable {
  query: Query
}
```

Directive:
```python
import logging

from typing import Any, Callable, Dict

from tartiflette.directive import CommonDirective, Directive

logger = logging.getLogger(__name__)


@Directive("nonIntrospectable")
class NonIntrospectable(CommonDirective):
    @staticmethod
    def on_introspection(
        _directive_args: Dict[str, Any],
        _next_directive: Callable,
        _introspected_element: Any,
    ) -> None:
        logger.debug(">> NonIntrospectable.on_introspection")
        return None
```

When querying, the log is never triggered.

### PR DESCRIPTION
**Pull request by abusi**
closes #134 

This PR make the SCHEMA directive location working through 2 new directive hooks 
  * for QUERY/MUTATION "on_schema_execution"
  * for Subscriptions "on_schema_subscriptions"

TODO:
  * [x] Code
  * [x] Tests
  * [x] Documentation
  * [x] Changelog and Release.

### CODE REVIEW

File: tartiflette/engine.py [hunk 1]
```diff
@@ -256,8 +258,75 @@ def __init__(
...
+            self._perform_query, self._perform_subsciption
+        )
+
         self._cooked = True
 
+    async def _perform_subsciption(
+        self,
+        schema,
+        document,
+        errors,
```
Comment (Maximilien-R) [inline tartiflette/engine.py]: I think that `errors` isn't enough descriptive on what this parameter could contains.

Maybe we could rename it `request_parsing_errors` or something more explicit? (same apply for `_perform_query` method)

What do you think?

File: tartiflette/utils/directives.py [hunk 2]
```diff
@@ -203,12 +210,14 @@ def wraps_with_directives(
     :param is_resolver: determines whether or not the wrapped func is a
     resolver
     :param with_default: determines whether or not if there is no directives
+    :param is_subscription: determines whether or not the wrapped func is a generator
```
Comment (Maximilien-R) [inline tartiflette/utils/directives.py]: Shouldn't it be named `is_async_generator` instead?

File: tartiflette/engine.py [hunk 3]
```diff
@@ -284,17 +353,16 @@ def __init__(
...
             self._schema,
             document,
-            self._build_response,
-            initial_value,
+            errors,
+            operation_name,
             context,
             variables,
-            operation_name,
+            initial_value,
```
Comment (Maximilien-R) [inline tartiflette/engine.py]: You should filled in the `context_coercer` parameter in order to forward it to potentials arguments directives:

```suggestion
            initial_value,
            context_coercer=context,
```

File: tartiflette/engine.py [hunk 4]
```diff
@@ -322,33 +390,17 @@ def __init__(
...
-            yield await execute(
-                self._schema,
-                document,
-                self._build_response,
-                payload,
-                context,
-                variables,
-                operation_name,
-            )
+            initial_value,
```
Comment (Maximilien-R) [inline tartiflette/engine.py]: You should filled in the `context_coercer` parameter in order to forward it to potentials arguments directives:

```suggestion
            initial_value,
            context_coercer=context,
```

File: tartiflette/utils/directives.py [hunk 5]
```diff
@@ -188,12 +205,19 @@
     return await resolver(*args, **kwargs)
 
 
+async def subscription_generator(generator: Callable, *args, **kwargs):
```
Comment (Maximilien-R) [inline tartiflette/utils/directives.py]:
```suggestion
async def subscription_generator(generator: AsyncGenerator, *args, **kwargs):
```

File: tartiflette/utils/directives.py [hunk 6]
```diff
@@ -175,6 +175,23 @@
     )
 
 
+async def directive_generator(
+    directive_func: Callable,
```
Comment (Maximilien-R) [inline tartiflette/utils/directives.py]:
```suggestion
    directive_func: AsyncGenerator,
```

File: tartiflette/schema/schema.py [hunk 7]
```diff
@@ -629,10 +643,14 @@ def _validate_directive_implementation(self) -> List[str]:
         for directive in self._directive_definitions.values():
             for expected in _IMPLEMENTABLE_DIRECTIVE_HOOKS:
                 attr = getattr(directive.implementation, expected, None)
-                if attr and not is_valid_coroutine(attr):
+                if (
+                    attr
+                    and not is_valid_coroutine(attr)
+                    and not is_valid_async_generator(attr)
```
Comment (Maximilien-R) [inline tartiflette/schema/schema.py]: Nones of directive hooks specified in `_IMPLEMENTABLE_DIRECTIVE_HOOKS` should be an `AsyncGenerator`. I'm not sure to understand this new check.

Comment (abusi) [reply tartiflette/schema/schema.py]: on_schema_subscription is an AsyncGenerator.

Comment (Maximilien-R) [reply tartiflette/schema/schema.py]: My bad, when I checked, I didn't see `on_schema_subscription` into the `_IMPLEMENTABLE_DIRECTIVE_HOOKS` constant 👍

In the future maybe we could add two checks, one for async generator and one for coroutine with two distincts whitelist constants (in order to make sure that we do have async generator and coroutine where they are expected).

Comment (abusi) [reply tartiflette/schema/schema.py]: Yep, good point i'll do it now, no point in waiting

**Pull request merged**